### PR TITLE
Pi Backend V2.3 (Heartbeat)

### PR DESCRIPTION
**Pull request by Jgerbrandt**
- Frontend Heartbeat Implantation
- Added Heartbeat Collection
- Backend sends heartbeat to DMB every 5 seconds.
- Abort is sent to DMB after 20 minutes.

### CODE REVIEW

File: backend/src/HeartbeatHandler.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    def start(self):
+        self.reset_timer()
+
+def heartbeat_thread(thread_name: str, heartbeat_workq: mp.Queue, message_handler_workq: mp.Queue):
+    """
+    The main loop of the heartbeat handler. This function will process the messages from the workq and spawn the dmb heartbeat thread.
+    """
+
+    # 20 minutes
+    timeout_seconds = 1200
```
Comment (cjchanx) [inline backend/src/HeartbeatHandler.py:54]: Note for later, we still gotta make Configuration.py have all the constants

File: backend/src/HeartbeatHandler.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    dmb_heartbeat_send_thread = threading.Thread(target=dmb_heartbeat_send, args=(heartbeat_handler,))   
+    dmb_heartbeat_send_thread.start()
+
+    while (1):
+        if not process_workq_message(heartbeat_workq.get(block=True), heartbeat_handler):
+            return
+
+def dmb_heartbeat_send(heartbeat_handler: HeartbeatHandler):
+    while heartbeat_handler.running:
```
Comment (cjchanx) [inline backend/src/HeartbeatHandler.py:66]: The thread will stop itself once the .runnning flag is set to False, fix would probably be to properly close down the thread on `.stop()` (.join) and then restart the thread in `.start()`.

Or change to `while True` and just have this as an if statement inside, downside to this one is there needs to be another mechanism to properly close the thread

**Review by cjchanx — APPROVED** [review]

File: backend/src/HeartbeatHandler.py (hunk 2 above)
Comment (Jgerbrandt) [reply backend/src/HeartbeatHandler.py]: Good catch. Done

**Review by EtSubas — COMMENTED** [review]
mostly nit

File: backend/src/HeartbeatHandler.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    def handle_timeout(self):
+        logger.error("Heartbeat timeout, aborting the DMB")
+        self.stop()
+        HeartbeatHandler.send_message_workq.put(
+            WorkQ_Message(
+                HeartbeatHandler.thread_name,
+                'radio',
+                THREAD_MESSAGE_SERIAL_WRITE,
+                ("RSC_ANY_TO_ABORT", "NODE_DMB", 0, 0,)
```
[the remark this answers is not in the document]
Comment (Jgerbrandt) [reply backend/src/HeartbeatHandler.py:37]: not with the DMB but it does show a serial command being sent in the logger

**Pull request merged**